constraint/pkg/client/matchers.go

```diff
@@ -0,0 +1,158 @@
+type constraintMatchers struct {
+	// matchers is the set of Constraint matchers by their Target, Kind, and Name.
+	// matchers is a map from Target to a map from Kind to a map from Name to matcher.
+	matchers map[string]map[string]map[string]constraints.Matcher
+
+	mtx sync.RWMutex
+}
+
+// Add inserts the Matcher for the Constraint with kind and name.
+// Replaces the current Matcher if one already exists.
+func (c *constraintMatchers) Add(key matcherKey, matcher constraints.Matcher) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if c.matchers == nil {
+		c.matchers = make(map[string]map[string]map[string]constraints.Matcher)
+	}
+
+	targetMatchers := c.matchers[key.target]
+	if targetMatchers == nil {
+		targetMatchers = make(map[string]map[string]constraints.Matcher)
+	}
+
+	kindMatchers := targetMatchers[key.kind]
+	if kindMatchers == nil {
+		kindMatchers = make(map[string]constraints.Matcher)
+	}
+
+	kindMatchers[key.name] = matcher
+	targetMatchers[key.kind] = kindMatchers
+	c.matchers[key.target] = targetMatchers
+}
+
+// Remove deletes the Matcher for the Constraint with kind and name.
+// Returns normally if no entry for the Constraint existed.
+func (c *constraintMatchers) Remove(key matcherKey) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if len(c.matchers) == 0 {
+		return
+	}
+
+	targetMatchers := c.matchers[key.target]
+	if len(targetMatchers) == 0 {
+		return
+	}
+
+	kindMatchers := targetMatchers[key.kind]
+	if len(kindMatchers) == 0 {
+		return
+	}
+
+	delete(kindMatchers, key.name)
+
+	// Remove empty parents to avoid memory leaks.
+	if len(kindMatchers) == 0 {
+		delete(targetMatchers, key.kind)
+	} else {
+		targetMatchers[key.kind] = kindMatchers
+	}
+
+	if len(targetMatchers) == 0 {
+		delete(c.matchers, key.target)
+	} else {
+		c.matchers[key.target] = targetMatchers
+	}
+}
+
+// RemoveAll removes all Matchers for Constraints with kind.
+// Returns normally if no entry for the kind existed.
+func (c *constraintMatchers) RemoveAll(kind string) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if len(c.matchers) == 0 {
+		return
+	}
+
+	for h, handlerMatchers := range c.matchers {
+		delete(handlerMatchers, kind)
+
+		if len(handlerMatchers) == 0 {
+			// It is safe to delete keys from a map while traversing it.
+			delete(c.matchers, h)
+		} else {
+			c.matchers[h] = handlerMatchers
+		}
+	}
+
+	delete(c.matchers, kind)
+}
+
+// ConstraintsFor returns the set of Constraints which should run against review
+// according to their Matchers. Returns a map from Kind to the names of the
+// Constraints of that Kind which should be run against review.
+//
+// Returns errors for each Constraint which was unable to properly run match
+// criteria.
+func (c *constraintMatchers) ConstraintsFor(review interface{}) (map[string]map[string][]string, error) {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	result := make(map[string]map[string][]string)
+
+	errs := errors.ErrorMap{}
+
+	for target, targetMatchers := range c.matchers {
```

Comment:
What happens if a target is broadcasting that it doesn't deal with a given review via `HandleReview`?

https://github.com/open-policy-agent/frameworks/blob/4ee2501883a584d60b5756dc69888c547e867ba1/constraint/pkg/handler/handler.go#L75-L83

Also, `HandleReview()` should be called to preprocess the review to ensure it's in an appropriate format for `Match` (ex. AugmentedReview in Gatekeeper adding the namespace).

Reply:
Done. Client now calls the matching logic per-target.